### PR TITLE
Added handling for a missing value of lbrel. Tests to follow.

### PR DESCRIPTION
**Pull request by gm-s**
This is a simple implementation to override lbrel when IMBI value is given. Setting to current latest value. 'Interactively tested', putting in some unit tests now - but feedback on my implementation would be welcome!

### CODE REVIEW

**Comment by pp-mo** [comment]
Hi @gm-s .  
Just starting to look into this.
If you're not familiar with the test_coding_standards.TestLicenseHeaders :  I think all it wants here is spaces around the '-' between the years.

**Comment by gm-s** [comment]
Thanks Patrick.

File: lib/iris/experimental/um.py [hunk 1]
```diff
@@ -520,7 +526,14 @@ def constants(name, dtype):
                     else:
                         offset = running_offset
                         data_provider = data_class(source, offset, word_size)
-                    klass = _FIELD_CLASSES[raw_headers[Field.LBREL_OFFSET]]
+                    lbrel_value = raw_headers[Field.LBREL_OFFSET]
+                    if lbrel_value == mdi_value:
```
Comment (pp-mo) [inline lib/iris/experimental/um.py]: (as previously discussed)
I think that it probably desirable to use the proper MDI value from the integer_constants section here (that is ```self.integer_constants[21]```).
Despite what I said before, from inspecting the PP code, I *don't* think we need to handle either the stated MDI _or_ the default value.

**Comment by pp-mo** [comment]
(As discussed offline)
The basically looks ok, *except* ...
We have long had much grief + concern over the use of warnings, because:
 * they are noisy
 * people mostly ignore them
 * the more noise, the more people ignore them

We have stated many times in the past our much-preferred 'Iris way' with such issues, which can be stated something like  "Never assume by default; have the user assert the assumption".

So, instead of warnings we prefer to have errors, with resolution strategies under user control.  Unfortunately, this tends to complicate APIs (especially modal keywords, which is another thing we really hate).
So we still have some warnings, like guessing missed bounds and missing orography, but these do make life noisy (see any testing output).

So this seems like a great place (in experimental) to trial an alternative approach.  Let's control the behaviour with a switch that the user must assert; but in a control object instead of a passed value; make it also support a context manager usage ("with" statement) like Iris.FUTURE.
 * pinch the code from iris.__init__.Future, to provide a threadsafe control object
 * retain the 'context' implementation, so we can use a 'with' block with it.
 * give it a property that controls the fix on/off (off by default, of course). 
 * define the singleton control object based on that, analagous to "Iris.FUTURE = Iris.Future()).
 * use control property to control the behaviour
 * when the fix is 'off', raise an **Exception** instead of a warning.

So, the code to fix a load problem this way may look something like this:
```
with um.LOAD_CONTROL(assume_missing_lbrel_is_v3=True):
    my_ff = um.FieldsFileVariant(filename)
```

File: lib/iris/experimental/um.py (hunk 1 above)
Comment (gm-s) [inline lib/iris/experimental/um.py]: I've removed the changes I was putting in to handle both, and I'm now pulling mdi_value from self.integer_constants[21].

:)

**Comment by pp-mo** [comment]
Sorry to miss you on phone @gm-s -- not important.
Ideas on 'control object' still rolling on a bit. Couple more suggestions:
 * make the fix control flexible, e.g. ```um.LOAD_CONTROL(for_missing_lbrel_use=3)```
 * make sure the error message explains how to engage the fix

**Comment by gm-s** [comment]
Thanks @pp-mo, completely agreed. My existing change does seem to have rather spectacularly broken the Travis tests - so I'm clearly not reading the IMDI value in the correct manner. Suggestions?

**Comment by pp-mo** [comment]
>  My existing change does seem to have rather spectacularly broken the Travis tests - so I'm clearly not reading the IMDI value in the correct manner. Suggestions?
It seems that the failing tests are the ones using a different kind of file (ancillary), for which the definitions of the integer_constants array is different. 
Meanwhile, the F3 document does not seem to admit that 'missing' values may appear anywhere in lookup data (let alone that the LBREL value might be 'missing') -- but the F3 document really only covers UM dumps (I really know nothing about ancillarys, of whatever sort).

On @marqh advice, I'll chase this up with the customer, so we can be sure what we need to support.
For the meantime, maybe we should revert to -32768.  Sorry, thought I knew what I was doing here..

**Comment by gm-s** [comment]
Not a problem, it's currently all of half-a-line of code in any case. I've changed it to pass the tests again for now.

**Comment by gm-s** [comment]
Hi @pp-mo - You mentioned on the phone about not changing the header - This means that we handle the field as a v3 field, but if anyone tests the version it still shows up as -32768.

```
>>> companion_filename = "iris_bad_lbrel.ff"
>>> from iris.experimental import um
>>> companion = um.FieldsFileVariant(companion_filename, mode=um.FieldsFileVariant.READ_MODE)
um.py:551: UserWarning: Encountered a field with a missing release value. Defaulting to 3.
  UserWarning)
>>> print companion.fields[0].lbrel
-32768
```
Does that cause any issues?

**Comment by pp-mo** [comment]
> >  My existing change does seem to have rather spectacularly broken the Travis tests...
> Maybe we will have revert to -32768.  

I've spoken to the customer (Steve Wardle) now, and his (their) view is that *ideally* the um.FieldsFileVariant interface needs to be as low-level, and make as few assumptions as is possible, so it can be used to 'fix' files that don't load into other software, for whatever reason.

In this particular case, he suggests that when we don't see an LBREL value we recognise, we should 'try' to interpret as version-3 if possible (and even then ideally be robust against a subsequent failure?).
So you can take the approach : just test for ```lbrel_value in valid_lbrels```, which I think will fix the immediate problem.
I think for now, I'm proposing we should test {{{lbrel_value not in _FIELD_CLASSES}}}.

To be clear, there isn't a clear scope for what we can and can not rely on for this code to work. I don't think the code we currently have is really designed with that in mind -- it would be nice if we had a separate, minimum-assumptions low level interface, but we are not starting from there.  We may even have to work towards that as we go.

**Comment by pp-mo** [comment]
> I think for now, I'm proposing we should test lbrel_value not in _FIELD_CLASSES

P.S. I also think we should leave the header value (i.e. "self.fields[xx].lbrel") unchanged, and not rewrite it.
That seems to be ok with the existing tests.

**Comment by rhattersley** [comment]
> In this particular case, he suggests that when we don't see an LBREL value we recognise, we should 'try' to interpret as version-3 if possible

:-1: for picking a particular version. It ensures confusion and/or deprecation complexities should there ever be a version four, and there are a lot of legacy files around encoded as version two.

I suggest we copy the behaviour for LBREL == -99 and just return an instance of the vanilla `Field` class. We can always add `Field`->`Field[N]` conversion methods if there's any real need.

> I'm proposing we should test `lbrel_value not in _FIELD_CLASSES`.

:+1: for using `_FIELD_CLASSES` as the reference information, but I'm not sure we need to "test" for anything. How about:
```python
klass = _FIELD_CLASSES.get(raw_headers[Field.LBREL_OFFSET], Field)
```

**Comment by gm-s** [comment]
I'm not sure how I could also raise a warning while using dict.get... I'd need to do a separate if statement to test for 'in' anyway.

If we do away with the warning as well, that's the simple option though.

**Comment by rhattersley** [comment]
> I'm not sure how I could also raise a warning while using dict.get

I'd not mentioned getting rid of the warning because @pp-mo had already suggested it, but since we're on the topic ... :+1: for getting rid of the warning. :grin:

**Comment by gm-s** [comment]
That's fine then. @pp-mo has spoken to the customer, so I think we can safely change the scope of the request. ;)

File: lib/iris/tests/unit/experimental/um/test_FieldsFileVariant.py [hunk 2]
```diff
@@ -75,6 +75,14 @@ def test(self):
         ffv = FieldsFileVariant(path)
         self.assertEqual(ffv.filename, path)
 
+@tests.skip_data
+class Test_class_assignment(tests.IrisTest):
+    def test(self):
+        path = tests.get_data_path(('FF', 'lbrel_test_data'))
+        ffv = FieldsFileVariant(path)
+        self.assertIs(ffv.fields[0], Field)
```
Comment (pp-mo) [inline lib/iris/tests/unit/experimental/um/test_FieldsFileVariant.py]: I think just to make it clear, I'd also check/assert the LBREL values, e.g. :
```
self.assertEqual(ffv.fields[0].int_headers[Field.LBREL_OFFSET], -32768)
self.assertEqual(ffv.fields[1].int_headers[Field.LBREL_OFFSET], 3)
```

**Comment by pp-mo** [comment]
Looking neat now (so much talk for something so small !)

Nearly there, but please consider previous comment, and fix it where pep8 wants 2 blank lines.

Meanwhile, I've merged your iris-test-data change now...
So, you need to update the reference in .travis.yml to run with the latest iris-test-data :
https://github.com/gm-s/iris/blob/f40dfc76ebfc8613dc620b60dcb067f20f8abf12/.travis.yml#L23

Then it should pass...

**Comment by gm-s** [comment]
@pp-mo @rhattersley I'm still not sure about my tests here, but I'm not sure of a more elegant way to do it. Any suggestions?

Cheers,
Gray.

**Comment by pp-mo** [comment]
> not sure about my tests here, but I'm not sure of a more elegant way

Testing it without a real file load looks it would make a lot of work, so as-is is fine by me.
In retrospect, this might have gone into the integration test instead, but looking at the existing code in there that horse is long gone !

**Comment by pp-mo** [comment]
Thanks @gm-s 
Sorry for the unexpected grief - it's been nit-pickin' good...

**Pull request merged**